server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2852,21 +2842,46 @@ public ParsedDocument newNoopTombstoneDoc(String reason) {
     void resetEngineToGlobalCheckpoint() throws IOException {
         assert getActiveOperationsCount() == 0 : "Ongoing writes [" + getActiveOperations() + "]";
         sync(); // persist the global checkpoint to disk
-        final long globalCheckpoint = getGlobalCheckpoint();
-        final Engine newEngine;
+        final SeqNoStats seqNoStats = seqNoStats();
+        final TranslogStats translogStats = translogStats();
+        // flush to make sure the latest commit, which will be opened by the read-only engine, includes all operations.
+        flush(new FlushRequest());
```

Comment:
do we need to set `waitIfOngoing` to be 100% sure it's finished?

Reply:
Yes, I pushed https://github.com/elastic/elasticsearch/pull/36159/commits/a907a8e4214281528575379a6d4b3c51256a0b39